codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/IdempotencyTokenMiddlewareGenerator.java

```diff
@@ -74,23 +105,47 @@ public void writeAdditionalFiles(
             SymbolProvider symbolProvider,
             TriConsumer<String, String, Consumer<GoWriter>> writerFactory
     ) {
-        ServiceShape serviceShape =  settings.getService(model);
+        ServiceShape serviceShape = settings.getService(model);
         Map<ShapeId, MemberShape> map = getOperationsWithIdempotencyToken(model, serviceShape);
         if (map.size() == 0) {
             return;
         }
 
-        writerFactory.accept("idempotencyTokenMiddleware.go", settings.getModuleName(), writer -> {
-            writer.write("var randReader io.Reader = cryptorand.Reader");
-            writer.addUseImports(SmithyGoDependency.IO);
-            for (Map.Entry<ShapeId, MemberShape> entry : map.entrySet()) {
-                ShapeId operationId = entry.getKey();
-                OperationShape operation = model.expectShape(operationId).asOperationShape().get();
-                Shape inputShape  = model.expectShape(operation.getInput().get());
-                MemberShape memberShape = entry.getValue();
-                execute(writer, symbolProvider, operation, memberShape, symbolProvider.toSymbol(inputShape));
-            }
+        writerFactory.accept("api_client.go", settings.getModuleName(), writer -> {
+            writer.write("// IdempotencyTokenProvider interface for providing idempotency token");
+            writer.openBlock("type IdempotencyTokenProvider interface {",
+                    "}", () -> {
+                        writer.write("GetToken() (string, error)");
+                    });
+            writer.write("");
         });
+
+        for (Map.Entry<ShapeId, MemberShape> entry : map.entrySet()) {
+            ShapeId operationShapeId = entry.getKey();
+            OperationShape operation = model.expectShape(operationShapeId, OperationShape.class);
+            writerFactory.accept(String.format("api_op_%s.go", operationShapeId.getName()), settings.getModuleName(),
```

Comment:
Would be better to use the GoDeledgator.useShapeTestWriter method instead of literal string. for file name.Especially since Operation shape is available.

```
delegator.useShapeWriter(operation, (writer) -> {
...
})
```